[정민석] RacingGame 3단계 리팩토링 리뷰 부탁드립니다.

dadadamarine opened the pull request:
개인적으로 OutputView.drawWinner() 메서드에서
racingGameController객체를 따로 생성해야 한다는 점이 마음에 안드네요.. 혹시 조언 주실분은 부탁드립니다!

dadadamarine committed:
add : 자동차 이름추가
1. inputView.getCarsNumber() 수정
2. model.Car 의 name 속성 추가 및 생성자 변경
3. RacingGameController.makeCars() 수정

dadadamarine committed:
add : 승자탐색 메서드들 추가.
1. RacingGame.getWinners , RacingGame.getMaxNumber, RacingGame.getMaxNumberCarNamesString, RacingGame.getNameIfHaveMaxNumber, RacingGame.getBiggerNumber 메서드 추가.
2.각 메서드의 테스팅소스 추가

dadadamarine committed:
add : 승자탐색 메서드들 추가.
1. RacingGame.getWinners , RacingGame.getMaxNumber, RacingGame.getMaxNumberCarNamesString, RacingGame.getNameIfHaveMaxNumber, RacingGame.getBiggerNumber 메서드 추가.
2.각 메서드의 테스팅소스 추가

dadadamarine committed:
fix : 오류수정 및 정렬
1. RacingGameController에서 moveCars메서드 수정 및 tryForTryNumber 메서드 추가
2. 정렬

wijehyeon commented on src/main/java/racingGame/view/OutputView.java:
```diff
@@ -1,21 +1,38 @@
+    }
+
+    public static void drawWinner(Car[] cars) {
+        RacingGameController racingGameController = new RacingGameController();
```
제생각으론 지금 우승자의 이름을 가져와야 하는 부분인데
레이싱 게임 컨트롤러에 그 부분을 다 구현해둔게 문제라고 생각이드네요
dto 패키지를 만들어서 RacingGameResultDTO를 활용해보세요.

esp2ar0 replied:
인자로 받으면 되지 않나요?

esp2ar0 commented on src/main/java/racingGame/controller/RacingGameController.java:
```diff
@@ -38,10 +54,45 @@ public void checkMove(Car car) {
+        return "";
+    }
+
+    public int getBiggerNumber(int originalNumber, int newNumber) {
```
Math 라이브러리에서 기본으로 제공해주는 메소드에 똑같은 일을 하는게 있습니다

esp2ar0 commented on src/main/java/racingGame/controller/RacingGameController.java:
```diff
@@ -38,10 +54,45 @@ public void checkMove(Car car) {
+        return maxNumber;
+    }
+
+    public String getMaxNumberCarNamesString(Car[] cars, int maxNumber) {
```
메소드 명으로 봤을 때 뜻하고자 했던 바를 이해하기 힘드네요
또한 String은 반환 타입으로 명시돼있어서 중복의 느낌이 드는군요

esp2ar0 commented on src/main/java/racingGame/controller/RacingGameController.java:
```diff
@@ -38,10 +54,45 @@ public void checkMove(Car car) {
+        return carNamesString.substring(0, carNamesString.length() - 2);
+    }
+
+    public String getNameIfHaveMaxNumber(Car car, int maxNumber) {
```
메소드 이름에서 나는 두가지 일을 하고 있다 라는 느낌이 드는군요
명확하지 않아 보여요

esp2ar0 commented on src/main/java/racingGame/controller/RacingGameController.java:
```diff
@@ -38,10 +54,45 @@ public void checkMove(Car car) {
+        for (int i = 0; i < cars.length; i++) {
+            carNamesString += getNameIfHaveMaxNumber(cars[i], maxNumber);
+        }
+        return carNamesString.substring(0, carNamesString.length() - 2);
```
substring을 사용한 것은 끝부분의 ", "를지워주기 위함인 것을 아래 메소드를 참고해서야 이해했습니다.
보완이 필요해 보여요

esp2ar0 commented:
step1 풀리퀘에서 리뷰 남겼던 내용도 반영해주세요~
머지 후에 혁진님께서 추가 리뷰도 남겼네요

또한 자바8의 스트림 람다 기능을 적용해보기 바랍니다

esp2ar0 commented on src/main/java/racingGame/view/OutputView.java:
```diff
@@ -1,21 +1,38 @@
     private static final char TRACE = '-';
 
+    public static void drawResults(Car[] cars) {
+        System.out.println("");
```
개행문자를 써주는 것도 좋을 것 같습니다

dadadamarine committed:
fix : racingCar 1차 리뷰반영

dadadamarine committed:
fix : racingCar 1차 리뷰반영
1. RacingGameClient 클래스 추가

dadadamarine committed:
fix : RacingGame 2차 리뷰반영
1. outputView에서 계산해주던 코드를(racingGame.getWinners() 를 컨트롤러로 넘기고, String만 outputView로 넘어오게 함.

dadadamarine committed:
fix : RacingGame 2차 리뷰반영
2. getWinnerIfHaveMaxNumber 메서드 이름 변경

dadadamarine committed:
fix : RacingGame 2차 리뷰반영
3. 테스트 코드 변경

dadadamarine committed:
fix : CarService 생성, 컨트롤러에서 메서드 분리

dadadamarine committed:
add : RacingGameResultService 생성
1. CarService에서 결과를 위한 메서드들 가져옴
2. 테스트 성공

dadadamarine committed:
add : CarService 테스트 성공 , RacingGameController 테스트 성공

dadadamarine committed:
fix : 정렬

hyukjin-lee commented on src/main/java/racingGame/util/RacingGameUtil.java:
```diff
@@ -0,0 +1,27 @@
+
+import java.util.Random;
+
+public class RacingGameUtil {
```
Util의 성격은, 프로젝트 전체적으로 사용될 수 있는 보조적인 성격입니다. 예를들어 StringUtil 같은.. 이 유틸이 하는 역할들은 대부분 각 객체로 위임해도 되는 책임과 역할들인 것 같습니다.

hyukjin-lee commented on src/main/java/racingGame/util/RacingGameUtil.java:
```diff
@@ -0,0 +1,27 @@
+    public static int getRandomNumber() {
+        Random random = new Random();
+        return random.nextInt(RANDOM_MAX);
+    }
```
이건 RandomGenerator  유틸로 만들어서 빼도 좋을듯 싶네요

hyukjin-lee commented on src/main/java/racingGame/util/RacingGameUtil.java:
```diff
@@ -0,0 +1,27 @@
+            return true;
+        }
+        return false;
+    }
```
isMove 함수가 util에 있어야할까요?

hyukjin-lee commented on src/main/java/racingGame/Service/CarService.java:
```diff
@@ -0,0 +1,30 @@
+
+        Arrays.asList(cars).stream()
+                .forEach(car -> move(car, RacingGameUtil.getRandomNumber()));
+    }
```
?? return 안해주는데 어떻게 쓴거지?

hyukjin-lee commented on src/main/java/racingGame/Service/CarService.java:
```diff
@@ -0,0 +1,30 @@
+        if (RacingGameUtil.isMove(randomNumber)) {
+            car.move();
+        }
+    }
```
service 단 함수들이 왜 다 static인가요?

hyukjin-lee commented on src/main/java/racingGame/Service/RacingGameResultService.java:
```diff
@@ -0,0 +1,36 @@
+    public static String getWinnerNames(Car[] cars, int maxNumber) {
+        String carNames = "";
+        for (int i = 0; i < cars.length; i++) {
+            carNames += getWinnerName(cars[i], maxNumber);
```
string builder랑 buffer 공부해서 사용해보세요

dadadamarine replied:
view는 출력만 하기 위해서, GameResultController에서 String을 만들도록 소스 수정하였습니다.
이번엔 버퍼를 반영하지 못했지만 버퍼의 효과를 내기위해서 String 으로 \n 까지 받아가며 축적한 후에 마지막에만 출력하도록 구현하였습니다.

hyukjin-lee commented on src/main/java/racingGame/Service/RacingGameResultService.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    public static String getWinnerName(Car car, int maxNumber) {
+        if (car.getPosition() == maxNumber) {
```
get 사용 금지

wijehyeon replied:
도메인에 대한 DTO를 사용해보세요
get을 줄이면서 다른 레이어들과 데이터를 주고 받게 해줍니다

hyukjin-lee commented on src/main/java/racingGame/client/RacingGameClient.java:
```diff
@@ -0,0 +1,14 @@
+public class RacingGameClient {
+
+    public static void main(String[] args) {
+        new RacingGameController().start(InputView.getCarsNames(), InputView.getTryNumber());
```
클라이언트에서 start만 수행하네요?
클라이언트 코드만 봐도 이 게임이 어떻게 수행되는지 전체적으로 파악 가능해야합니다

hyukjin-lee reviewed (commented):
리뷰를 많이 했던 것 같은데 기성이나 너나 계속 엇나가는 방향으로 쭉 가는 느낌이네.. qna를 먼저해서 그런가? 여기서는 MVC 참고만 하고 책임과 역할에 따라서 클래스 분리하는 것만 하라고 했었는데 요구사항에 비해 너무 큰 layered아키텍처를 적용해서 전체적으로 가독성도 좀 떨어지고 효율성도 떨어지는듯. 상은이, 나, 창환이, 화중이형 소스 쭉 읽어봐

wijehyeon commented on src/main/java/racingGame/view/OutputView.java:
```diff
@@ -4,18 +4,30 @@
 public class OutputView {
     private static final char TRACE = '-';
+    private static final String RESULT_MESSAGE="\n실행결과\n";
+
```
OutputView인데
로직이 많이 들어가있는 것 같습니다.
View는 그저 보여주는 역할에 충실해야할 것 같습니다

wijehyeon reviewed (commented):
Car에 대한 domain만 생각해서 game의 정보를 담는 것들을 분산시켜 구현되어있는 느낌입니다
GameResult같은 도메인을 설정하면 여기저기 흩어진게 정리되면서 객체지향적인 느낌이 되지않을까?(추상적이라 죄송합니다)
자바8을 사용하려는시도는 좋습니다
